channels/sshagent/client/sshagent_main.c

```diff
@@ -0,0 +1,398 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * sshagent_main.c: DVC plugin to forward queries from RDP to the ssh-agent
+ *
+ * This relays data to and from an ssh-agent program equivalent running on the
+ * RDP server to an ssh-agent running locally.  Unlike the normal ssh-agent,
+ * which sends data over an SSH channel, the data is send over an RDP dynamic
+ * virtual channel.
+ *
+ * protocol specification:
+ *     Forward data verbatim over RDP dynamic virtual channel named "sshagent"
+ *     between a ssh client on the xrdp server and the real ssh-agent where
+ *     the RDP client is running.  Each connection by a separate client to
+ *     xrdp-ssh-agent gets a separate DVC invocation.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <pwd.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include <winpr/crt.h>
+#include <winpr/synch.h>
+#include <winpr/thread.h>
+#include <winpr/stream.h>
+
+#include "sshagent_main.h"
+#include <freerdp/channels/log.h>
+
+#define TAG CHANNELS_TAG("sshagent.client")
+
+typedef struct _SSHAGENT_LISTENER_CALLBACK SSHAGENT_LISTENER_CALLBACK;
+struct _SSHAGENT_LISTENER_CALLBACK
+{
+	IWTSListenerCallback iface;
+
+	IWTSPlugin* plugin;
+	IWTSVirtualChannelManager* channel_mgr;
+
+        const char *agent_uds_path;
+};
+
+typedef struct _SSHAGENT_CHANNEL_CALLBACK SSHAGENT_CHANNEL_CALLBACK;
+struct _SSHAGENT_CHANNEL_CALLBACK
+{
+	IWTSVirtualChannelCallback iface;
+
+	IWTSPlugin* plugin;
+	IWTSVirtualChannelManager* channel_mgr;
+	IWTSVirtualChannel* channel;
+
+        int agent_fd;
+	HANDLE thread;
+	CRITICAL_SECTION lock;
+};
+
+typedef struct _SSHAGENT_PLUGIN SSHAGENT_PLUGIN;
+struct _SSHAGENT_PLUGIN
+{
+	IWTSPlugin iface;
+
+	SSHAGENT_LISTENER_CALLBACK* listener_callback;
+};
+
+
+/**
+ * Function to open the connection to the sshagent
+ *
+ * @return The fd on success, otherwise -1
+ */
+static int connect_to_sshagent(const char *udspath)
+{
+        int agent_fd = socket(AF_UNIX, SOCK_STREAM, 0);
+        if (agent_fd == -1)
+        {
+                WLog_ERR(TAG, "Can't open Unix domain socket!");
+                return -1;
+        }
+
+        struct sockaddr_un addr;
+        memset(&addr, 0, sizeof(addr));
+        addr.sun_family = AF_UNIX;
+        strncpy(addr.sun_path, udspath, sizeof(addr.sun_path) - 1);
+        int rc = connect(agent_fd, (struct sockaddr*)&addr, sizeof(addr));
+        if (rc != 0)
+        {
+                WLog_ERR(TAG, "Can't connect to Unix domain socket \"%s\"!",
```

Comment:
agent_fd is leaked here

Reply:
... when connect() fails.  Thanks - fixed.